### PR TITLE
next gen build tasks

### PR DESCRIPTION
**Pull request by CITguy**
Updated build tasks to accommodate `src/components/*`

### LGTMs
- [x] Dev LGTM
- [x] Dev LGTM
- [x] Dev LGTM

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -23,5 +23,12 @@
     "angular-mocks": "1.2.28",
     "jquery": "2.0.3",
     "pure": "0.5.0"
+  },
+  "overrides": {
+      "font-awesome": {
+          "main": [
+              "css/font-awesome.css"
+          ]
+      }
```
Comment (CITguy) [inline bower.json]: This fixes the issue with font-awesome disappearing in `demo/index.html`

Comment (bry4n) [inline bower.json]: woot!

File: demo/components.html [hunk 2]
```diff
@@ -1,42 +1,37 @@
...
+          <td>
+            <a href="#/components/{{comp.name}}">{{ comp.name }}</a><br />
+          </td>
+          <td>
+            <span ng-if="comp.stability">
+              <img ng-src="http://badges.github.io/stability-badges/dist/{{comp.stability}}.svg"></img>
+            </span>
+          </td>
+        </tr>
+      </tbody>
```
Comment (CITguy) [inline demo/components.html]: Mostly reindented. After speaking with @lfielder, removed `Last Updated` column until IA stabilizes and we can find a better way to present that information.

File: demo/index.html [hunk 3]
```diff
@@ -33,9 +33,9 @@
 
     <script src="dist/<%= config.fileNameTpl %>.js"></script>
     <script src="demo.js"></script>
-    <script src="components.js"></script>
+    <script src="modules.js"></script>
```
Comment (CITguy) [inline demo/index.html]: new `modules.js` will contain ALL modules across quarks, atoms, molecules, etc.

File: demo/modules.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+angular.module('demoApp')
+.value('components', <%= JSON.stringify(config.components, null, 4) %>)
```
Comment (CITguy) [inline demo/modules.js]: this value is likely to go away, considering that we can filter `Modules` (below) by their `category` property (`quarks`, `atoms`, `molecules`, `components`)

File: grunt-tasks/modules.js [hunk 5]
```diff
@@ -10,221 +11,118 @@ module.exports = function (grunt) {
...
+        var _docHtmlFiles = grunt.file.expand(globBase + '/docs/*.html');
+        var _docLessFiles = grunt.file.expand(globBase + '/*.less');
+        var _docMidwayFiles = grunt.file.expand(globBase + '/docs/*.midway.js');
+        var _docExerciseFiles = grunt.file.expand(globBase + '/**/*.exercise.js');
+
+        // Fetch Module Metadata
+        var metadataJson = globBase + '/' + name + '.meta.json';
+        var metadata = {};
+        if (grunt.file.exists(metadataJson)) {
+            metadata = grunt.file.readJSON(metadataJson);
```
Comment (CITguy) [inline grunt-tasks/modules.js]: Use `*.meta.json` (if present) in each component directory as a base for information needed in this task.

File: grunt-tasks/modules.js [hunk 6]
```diff
@@ -10,221 +11,118 @@ module.exports = function (grunt) {
...
+                ].join('\n')
             }
-        };
+        });
 
-        module.dependencies.forEach(findModule);
         grunt.config('config.modules', grunt.config('config.modules').concat(module));
-    }
-
-    function dependenciesForModule (name) {
```
Comment (CITguy) [inline grunt-tasks/modules.js]: function is unnecessary considering that all dependencies are captured on first pass

File: grunt-tasks/modules.js [hunk 7]
```diff
@@ -10,221 +11,118 @@ module.exports = function (grunt) {
...
 
     grunt.registerTask('modules', 'Auto-discover EncoreUI modules and load their details into config', function () {
-        var _ = grunt.util._;
-
-        moduleMetaData();
-
-        grunt.file.expand({ filter: 'isDirectory', cwd: '.' }, 'src/!(styles)')
+        // Iterate over source modules to build config.modules
+        grunt.file
+            .expand({ filter: 'isDirectory', cwd: '.' }, 'src/*/!(styles)')
```
Comment (CITguy) [inline grunt-tasks/modules.js]: note to self: file glob is incorrect (`styles` is still at `src/styles`)

File: grunt-tasks/modules.js [hunk 8]
```diff
@@ -10,221 +11,118 @@ module.exports = function (grunt) {
...
-        var modules = grunt.config('config.modules');
-        // remove any modules that don't have javascript
+        /* sorted modules array for various filtering */
+        var modules = _.sortBy(grunt.config('config.modules'), 'name');
+
+        /* Fetch Atoms, Components, Molecules, and Quarks from all modules */
+        grunt.config('config.atoms', _.filter(modules, { category: 'atoms' }));
+        grunt.config('config.components', _.filter(modules, { category: 'components' }));
+        grunt.config('config.molecules', _.filter(modules, { category: 'molecules' }));
+        grunt.config('config.quarks', _.filter(modules, { category: 'quarks' }));
```
Comment (CITguy) [inline grunt-tasks/modules.js]: `config.components` is the only one being used at the moment, but the other three will soon be needed.

File: grunt-tasks/modules.js [hunk 9]
```diff
@@ -10,221 +11,118 @@ module.exports = function (grunt) {
...
         // Set the concat task to concatenate the given src modules
+        var srcFiles = _.pluck(modules, 'srcFiles'); // OK
         grunt.config('concat.dist.src', grunt.config('concat.dist.src').concat(srcFiles));
 
         // Set the concat-with-templates task to concat the given src & tpl modules
+        var tplJsFiles = _.pluck(modules, 'tplJsFiles'); // OK
         grunt.config('concat.distTpls.src', grunt.config('concat.distTpls.src').concat(srcFiles).concat(tplJsFiles));
-
-        // Set the copy task to process via grunt template
-        grunt.config('copy.demohtml.options.process', grunt.template.process);
```
Comment (CITguy) [inline grunt-tasks/modules.js]: this is the only place modifying `copy.demohtml.options.process`, so the configuration was set explicitly in `grunt-tasks/options/copy.js`

File: grunt-tasks/options/complexity.js [hunk 10]
```diff
@@ -1,10 +1,9 @@
 module.exports = {
     generic: {
         src: [
-            'src/*/*.js',
-            '!src/*/*.spec.js',
-            '!src/*/*.page.js',
-            '!src/configs/configs.js'
+            'src/**/*.js',
+            '!src/**/*.spec.js',
+            '!src/**/*.page.js'
```
Comment (CITguy) [inline grunt-tasks/options/complexity.js]: updated globs for new file structure

File: grunt-tasks/options/karma.js [hunk 11]
```diff
@@ -24,9 +24,9 @@ module.exports = {
     options: {
         configFile: 'karma.conf.js',
         files: files.concat([
-            'src/*/*.js',               // src/<component>/<component>.js
-            'src/*/templates/*.html',   // src/<component>/templates/<item>.html
-            'src/*/scripts/*.js'        // src/<component>/scripts/<item>.js
+            'src/**/*.module.js',       // src/:category/:component/<component>.module.js
```
Comment (CITguy) [inline grunt-tasks/options/karma.js]: This depends on remaining decomposition work to normalize all components to `<component>.module.js`

File: grunt-tasks/options/replace.js [hunk 12]
```diff
@@ -11,29 +13,15 @@ module.exports = {
...
+            'bower/*.map',
+            'bower/*.js'
+        ],
         overwrite: true,
         replacements: [{
             from: config.versionRegEx,
             to: ''
         }]
-    },
-    'wraith': {
```
Comment (CITguy) [inline grunt-tasks/options/replace.js]: Leftover config. Wraith is no longer being used.

File: src/components/grid/README.md [hunk 13]
```diff
@@ -1,5 +1,3 @@
-[![deprecated](http://badges.github.io/stability-badges/dist/deprecated.svg)](http://github.com/badges/stability-badges)
-
```
Comment (CITguy) [inline src/components/grid/README.md]: All component `README.md` files have had their stability badge removed. It is now being handled via the `*.meta.json` and updated `demo/component-template.html`

screen shot 2015-11-18 at 12 01 07 pm

File: src/components/rxApp/docs/rxApp.js [hunk 14]
```diff
@@ -55,7 +55,11 @@ angular.module('demoApp')
         }
     });
 
-    var searchDirective = 'rx-app-search placeholder="Enter User" model="$root.user" pattern="/^([0-9a-zA-Z._ -]{2,})$/"';
```
Comment (CITguy) [inline src/components/rxApp/docs/rxApp.js]: One of many lines of source code completely missed by incorrect file globbing with regards to jshint.

File: src/components/rxBreadcrumbs/scripts/rxBreadcrumbs.spec.js [hunk 15]
```diff
@@ -19,16 +19,6 @@ describe('Breadcrumbs', function () {
...
-    var defaultBreadcrumb = {
-        path: '/',
-        name: 'Home'
-    };
-
-    var newHome = {
-        path: '//foo/bar',
-        name: 'New Home'
-    };
-
```
Comment (CITguy) [inline src/components/rxBreadcrumbs/scripts/rxBreadcrumbs.spec.js]: unused variables missed by bad jshint file glob

File: src/components/rxBulkSelect/scripts/rxBulkSelectController.spec.js [hunk 16]
```diff
@@ -3,41 +3,6 @@
 describe('rxBulkSelect', function () {
     var scope, compile, rootScope, timeout;
 
-    var template =
```
Comment (CITguy) [inline src/components/rxBulkSelect/scripts/rxBulkSelectController.spec.js]: unused variable

File: src/components/rxAuth/docs/rxAuth.js [hunk 17]
```diff
@@ -1,10 +1,10 @@
 angular.module('demoApp')
 .controller('rxAuthCtrl', function ($scope, Auth) {
     $scope.hasRole = function () {
-        alert('Has "superhero" Role? : ' + Auth.hasRole('superhero'));
+        window.alert('Has "superhero" Role? : ' + Auth.hasRole('superhero'));
```
Comment (parlarjb) [inline src/components/rxAuth/docs/rxAuth.js]: Can we make this `$window`?

File: src/components/rxAuth/docs/rxAuth.js [hunk 18]
```diff
@@ -1,10 +1,10 @@
...
     };
 
     $scope.isAuthenticated = function () {
-        alert('Is Authenticated? : ' + Auth.isAuthenticated());
+        window.alert('Is Authenticated? : ' + Auth.isAuthenticated());
```
Comment (parlarjb) [inline src/components/rxAuth/docs/rxAuth.js]: Same here

**Comment by 100stacks** [comment]
@CITguy great catch with the next gen build tasks :+1:

File: src/components/rxAuth/docs/rxAuth.js (hunk 17 above)
Comment (CITguy) [inline src/components/rxAuth/docs/rxAuth.js]: sure can

File: grunt-tasks/modules.js [hunk 19]
```diff
@@ -9,222 +10,127 @@ module.exports = function (grunt) {
 
     //Common encore.ui module containing all modules for src and templates
     //findModule: Adds a given module to config
-    var foundModules = {};
```
Comment (CITguy) [inline grunt-tasks/modules.js]: variable is unnecessary

**Comment by bry4n** [comment]
LGTM

**Comment by 100stacks** [comment]
LGTM

**Comment by parlarjb** [comment]
LGTM

**Pull request merged**